script.service.ltvpl/resources/PL_Server.py

```diff
@@ -0,0 +1,477 @@
+        self.server=Server(address, maxConnections=5)
+        self.server.addDataReceivedEventHandler(self.onServerDataReceived)
+        
+        self.cmds={Cmd.AddPlayListItem:self.AddPlayListItem,
+                   Cmd.RemovePlayListItem:self.RemovePlayListItem,
+                   Cmd.RemoveAllPlayListItems:self.RemoveAllPlayListItems,
+                   Cmd.GetChPlayListItems:self.GetChPlayListItems,
+                   Cmd.GetPlayList:self.GetPlayList,
+                   Cmd.GetChGroupList:self.GetChGroupList,
+                   Cmd.GetChannelList:self.GetChannelList,
+                   Cmd.SkipEvent:self.SkipEvent,
+                   Cmd.GetPlayListItem:self.GetPlayListItem,
+                   Cmd.UpdatePlayListItem:self.UpdatePlayListItem,
+                   Cmd.SetVacationMode:self.SetVacationMode,
+                   Cmd.GetVacationMode:self.GetVacationMode,
+                   Cmd.SetPreRollTime:self.SetPreRollTime,
+                   Cmd.GetPreRollTime:self.GetPreRollTime,
+                   Cmd.EnablePlayListItem:self.EnablePlayListItem,
+                   Cmd.GetPlayListItemState:self.GetPlayListItemState,
+                   Cmd.DisablePlayListItem:self.DisablePlayListItem,
+                   Cmd.ClearPlayList:self.ClearPlayList}
+
+        DbgPrint("Server Initialized",MODULEDEBUGMODE=MODULEDEBUGMODE)
+
+
+    def addSettingsChangedEventHandler(self, handler):
+        self.dataSet.SettingsChangedEventHandler(handler)
+
+    def onDataSetEvent(self, data):
+        self.server.sendNotification(data)
+
+    @TS_decorator
+    def dailyMaintenance(self):
+        from time import sleep
+        """
+            Wait for EPG to be present before verifying dataset and
+            also before launching daily maintenance service 
+        """
+        count = 60
+        lastCh = lastChNumber = self.dataSet.LastChNumber
+
+        while count >= 0:
+            try:
+                if lastChNumber is None:
+                    lastCh = getLastChannelInfo(kodiObj)['channelnumber']
+                    if lastChNumber is None or lastCh > lastChNumber:
+                        lastChNumber = self.dataSet.LastChNumber = lastCh
+
+                if lastCh == lastChNumber and TvGuideIsPresent(kodiObj, lastCh):
+                    DbgPrint("Got Last Channel & Tv Guide...")
+                    self.dataSet.verifyDataset()
+                    break
+            except: pass
+
+            sleep(5)
+            count -= 1
+
+
+        DbgPrint("Daily Maintenance Operation Started...")
+
+        while self.dailyMaintenanceFlag:
+            today = datetime.today().date()
+            tomorrow = datetime.combine(today, datetime.min.time()) + timedelta(days=1, seconds=60)
+            seconds = tomorrow - datetime.now()
+            self.dailyMaintThread = Timer(seconds.total_seconds(), self.dataSet.verifyDataset)
+            self.dailyMaintThread.start()
+            time.sleep(1.0)
```

Comment:
use xbmc.sleep or xbmc.Monitor().waitForAbort

Reply:
This code also runs in a non-Kodi environment which is why time.sleep() is used.